File: src/main/java/no/nav/syfo/services/TilgangService.java
package no.nav.syfo.services;

import no.nav.syfo.axsys.AxsysConsumer;
import no.nav.syfo.domain.PersonInfo;
import no.nav.syfo.domain.Tilgang;
import no.nav.syfo.skjermedepersoner.SkjermedePersonerPipConsumer;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;

import static no.nav.syfo.config.CacheConfig.*;
import static no.nav.syfo.domain.AdRoller.*;

@Service
public class TilgangService {

    public static final String GEOGRAFISK = "GEOGRAFISK";

    private final AxsysConsumer axsysConsumer;
    private final LdapService ldapService;
    private final SkjermedePersonerPipConsumer skjermedePersonerPipConsumer;
    private final GeografiskTilgangService geografiskTilgangService;
    private final PersonService personService;

    private final static String ENHET = "ENHET";
    private final static String DISKRESJONSKODE_KODE6 = "SPSF";
    private final static String DISKRESJONSKODE_KODE7 = "SPFO";

    @Autowired
    public TilgangService(
            AxsysConsumer axsysConsumer,
            LdapService ldapService,
            SkjermedePersonerPipConsumer skjermedePersonerPipConsumer,
            GeografiskTilgangService geografiskTilgangService,
            PersonService personService
    ) {
        this.ldapService = ldapService;
        this.skjermedePersonerPipConsumer = skjermedePersonerPipConsumer;
        this.geografiskTilgangService = geografiskTilgangService;
        this.axsysConsumer = axsysConsumer;
        this.personService = personService;
    }

    public Tilgang sjekkTilgangTilBruker(String veilederId, String fnr) {
        PersonInfo personInfo = personService.hentPersonInfo(fnr);
        return sjekkTilgang(fnr, veilederId, personInfo);
    }

    @Cacheable(cacheNames = TILGANGTILBRUKER, key = "#veilederId.concat(#brukerFnr)", condition = "#brukerFnr != null && #veilederId != null")
    public Tilgang sjekkTilgang(String brukerFnr, String veilederId, PersonInfo personInfo) {
        if (!harTilgangTilTjenesten(veilederId)) {
            return new Tilgang().withHarTilgang(false).withBegrunnelse(SYFO.name());
        }

        if (!geografiskTilgangService.harGeografiskTilgang(veilederId, brukerFnr, personInfo.getGeografiskTilknytning())) {
            return new Tilgang().withHarTilgang(false).withBegrunnelse(GEOGRAFISK);
        }

        String diskresjonskode = personInfo.getDiskresjonskode();
        if (DISKRESJONSKODE_KODE6.equals(diskresjonskode)) {
            return new Tilgang().withHarTilgang(false).withBegrunnelse(KODE6.name());
        } else if (DISKRESJONSKODE_KODE7.equals(diskresjonskode) && !harTilgangTilKode7(veilederId)) {
            return new Tilgang().withHarTilgang(false).withBegrunnelse(KODE7.name());
        }

        if (skjermedePersonerPipConsumer.erSkjermet(brukerFnr) && !harTilgangTilEgenAnsatt(veilederId)) {
            return new Tilgang().withHarTilgang(false).withBegrunnelse(EGEN_ANSATT.name());
        }

        return new Tilgang().withHarTilgang(true);
    }

    @Cacheable(cacheNames = TILGANGTILTJENESTEN, key = "#veilederId", condition = "#veilederId != null")
    public Tilgang sjekkTilgangTilTjenesten(String veilederId) {
        if (harTilgangTilTjenesten(veilederId))
            return new Tilgang().withHarTilgang(true);
        return new Tilgang().withHarTilgang(false).withBegrunnelse(SYFO.name());
    }

    @Cacheable(cacheNames = TILGANGTILENHET, key = "#veilederId.concat(#enhet)", condition = "#enhet != null && #veilederId != null")
    public Tilgang sjekkTilgangTilEnhet(String veilederId, String enhet) {
        if (!harTilgangTilSykefravaersoppfoelging(veilederId))
            return new Tilgang().withHarTilgang(false).withBegrunnelse(SYFO.name());
        if (!harTilgangTilEnhet(veilederId, enhet))
            return new Tilgang().withHarTilgang(false).withBegrunnelse(ENHET);
        return new Tilgang().withHarTilgang(true);
    }

    private boolean harTilgangTilTjenesten(String veilederId) {
        return harTilgangTilSykefravaersoppfoelging(veilederId);
    }

    private boolean harTilgangTilSykefravaersoppfoelging(String veilederId) {
        return ldapService.harTilgang(veilederId, SYFO.rolle);
    }

    private boolean harTilgangTilKode7(String veilederId) {
        return ldapService.harTilgang(veilederId, KODE7.rolle);
    }

    private boolean harTilgangTilEgenAnsatt(String veilederId) {
        return ldapService.harTilgang(veilederId, EGEN_ANSATT.rolle);
    }

    private boolean harTilgangTilEnhet(String veilederId, String navEnhetId) {
        return axsysConsumer.enheter(veilederId)
                .stream()
                .anyMatch(enhet -> enhet.getEnhetId().equals(navEnhetId));
    }
}

File: src/main/kotlin/no/nav/syfo/tilgang/TilgangService.kt
package no.nav.syfo.tilgang

import io.micrometer.core.annotation.Timed
import no.nav.syfo.cache.CacheConfig
import no.nav.syfo.consumer.axsys.AxsysConsumer
import no.nav.syfo.consumer.graphapi.GraphApiConsumer
import no.nav.syfo.consumer.pdl.PdlConsumer
import no.nav.syfo.consumer.skjermedepersoner.SkjermedePersonerPipConsumer
import no.nav.syfo.domain.AdRoller
import no.nav.syfo.domain.PersonIdentNumber
import no.nav.syfo.geografisktilknytning.GeografiskTilgangService
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Service

@Service
class TilgangService @Autowired constructor(
    private val adRoller: AdRoller,
    private val axsysConsumer: AxsysConsumer,
    private val papirsykmeldingTilgangService: PapirsykmeldingTilgangService,
    private val graphApiConsumer: GraphApiConsumer,
    private val skjermedePersonerPipConsumer: SkjermedePersonerPipConsumer,
    private val geografiskTilgangService: GeografiskTilgangService,
    private val pdlConsumer: PdlConsumer
) {
    fun sjekkTilgangTilBruker(
        veilederId: String,
        fnr: String,
        consumerClientId: String = ""
    ): Tilgang {
        return sjekkTilgang(fnr, veilederId, consumerClientId)
    }

    @Cacheable(
        cacheNames = [CacheConfig.TILGANGTILBRUKER],
        key = "#veilederId.concat(#brukerFnr).concat(#consumerClientId)",
        condition = "#brukerFnr != null && #veilederId != null && #consumerClientId != null"
    )
    fun sjekkTilgang(
        brukerFnr: String,
        veilederId: String,
        consumerClientId: String
    ): Tilgang {
        if (!harTilgangTilTjenesten(veilederId)) {
            return Tilgang(
                harTilgang = false,
            )
        }
        if (!geografiskTilgangService.harGeografiskTilgang(veilederId, brukerFnr)) {
            return Tilgang(
                harTilgang = false,
            )
        }
        val personIdentNumber = PersonIdentNumber(brukerFnr)
        val pdlPerson = pdlConsumer.person(personIdentNumber.value)
        if (pdlConsumer.isKode6(pdlPerson) && !harTilgangTilKode6(veilederId)) {
            return Tilgang(
                harTilgang = false,
            )
        } else if (pdlConsumer.isKode7(pdlPerson) && !harTilgangTilKode7(veilederId)) {
            return Tilgang(
                harTilgang = false,
            )
        }
        return if (skjermedePersonerPipConsumer.erSkjermet(brukerFnr) && !harTilgangTilEgenAnsatt(veilederId)) {
            return Tilgang(
                harTilgang = false,
            )
        } else Tilgang(harTilgang = true)
    }

    @Cacheable(
        cacheNames = [CacheConfig.TILGANGTILBRUKER_PAPIRSYKMELDING],
        key = "#veilederIdent.concat(#personIdent).concat(#consumerClientId)",
        condition = "#personIdent != null && #veilederIdent != null && #consumerClientId !=null"
    )
    fun sjekkTilgangTilBrukerMedPapirsykmelding(
        consumerClientId: String,
        personIdent: String,
        veilederIdent: String,
    ): Tilgang {
        return if (papirsykmeldingTilgangService.harPapirsykmeldingTilgang(veilederIdent = veilederIdent)) {
            sjekkTilgang(
                consumerClientId = consumerClientId,
                brukerFnr = personIdent,
                veilederId = veilederIdent,
            )
        } else Tilgang(
            harTilgang = false,
        )
    }

    @Cacheable(cacheNames = [CacheConfig.TILGANGTILTJENESTEN], key = "#veilederId", condition = "#veilederId != null")
    fun sjekkTilgangTilTjenesten(veilederId: String): Tilgang {
        return if (harTilgangTilTjenesten(veilederId)) Tilgang(
            harTilgang = true
        ) else Tilgang(
            harTilgang = false,
        )
    }

    @Cacheable(cacheNames = [CacheConfig.TILGANGTILENHET], key = "#veilederId.concat(#enhet)", condition = "#enhet != null && #veilederId != null")
    fun sjekkTilgangTilEnhet(veilederId: String, enhet: String): Tilgang {
        if (!harTilgangTilSykefravaersoppfoelging(veilederId)) return Tilgang(
            harTilgang = false,
        )
        return if (!harTilgangTilEnhet(veilederId, enhet)) Tilgang(
            harTilgang = false,
        ) else Tilgang(
            harTilgang = true
        )
    }

    fun preloadCacheForBrukere(personIdentList: List<String>) {
        personIdentList.forEach { personIdent ->
            try {
                pdlConsumer.person(personIdent)
                pdlConsumer.geografiskTilknytningResponse(personIdent)
                skjermedePersonerPipConsumer.erSkjermet(personIdent, systemToken = true)
            } catch (exc: Exception) {
                log.warn("Failed to preload cache", exc)
            }
        }
    }

    @Timed("syfotilgangskontroll_harTilgangTilTjenesten", histogram = true)
    private fun harTilgangTilTjenesten(veilederId: String): Boolean {
        return harTilgangTilSykefravaersoppfoelging(veilederId)
    }

    private fun harTilgangTilSykefravaersoppfoelging(veilederId: String): Boolean {
        return graphApiConsumer.hasAccess(veilederId, adRoller.SYFO)
    }

    @Timed("syfotilgangskontroll_kode6TilgangSevice_harTilgang", histogram = true)
    fun harTilgangTilKode6(veilederId: String): Boolean {
        return graphApiConsumer.hasAccess(veilederId, adRoller.KODE6)
    }

    @Timed("syfotilgangskontroll_harTilgangTilKode7", histogram = true)
    private fun harTilgangTilKode7(veilederId: String): Boolean {
        return graphApiConsumer.hasAccess(veilederId, adRoller.KODE7)
    }

    @Timed("syfotilgangskontroll_harTilgangTilEgenAnsatt", histogram = true)
    private fun harTilgangTilEgenAnsatt(veilederId: String): Boolean {
        return graphApiConsumer.hasAccess(veilederId, adRoller.EGEN_ANSATT)
    }

    private fun harTilgangTilEnhet(veilederId: String, navEnhetId: String): Boolean {
        val veiledersEnheter = axsysConsumer.enheter(veilederId)
        val veilederHasAccessToEnhet = veiledersEnheter
            .stream()
            .anyMatch { (enhetId) -> enhetId == navEnhetId }

        if (!veilederHasAccessToEnhet) {
            val veiledersEnheterAsString = veiledersEnheter.joinToString(", ")
            log.info("Veileder har ikke tilgang til ønsket enhet! Ønsket enhet: $navEnhetId, veileders enheter: $veiledersEnheterAsString")
            return false
        }
        return true
    }

    companion object {
        const val GEOGRAFISK = "GEOGRAFISK"
        const val ENHET = "ENHET"
        private val log = LoggerFactory.getLogger(TilgangService::class.java)
    }
}
